src/intrusive_double_linked_list.rs

```diff
@@ -48,278 +50,335 @@ impl<T> DerefMut for ListNode<T> {
         let mut current = self.head;
 
-        while !current.is_null() {
-            if (*item).deref() < (*current).deref() {
-                // Need to insert before the current item
-                (*current).prev = item;
-                if !prev.is_null() {
-                    (*prev).next = item;
-                } else {
-                    // We are inserting at the beginning of the list
-                    self.head = item;
+        while let Some(mut current_node) = current {
+            if node.data < current_node.as_ref().data {
+                // Need to insert before the current node
+                current_node.as_mut().prev = Some(node.into());
+                match prev {
+                    Some(mut prev) => {
+                        prev.as_mut().next = Some(node.into());
+                    }
+                    None => {
+                        // We are inserting at the beginning of the list
+                        self.head = Some(node.into());
+                    }
                 }
-                (*item).next = current;
-                (*item).prev = prev;
+                node.next = current;
+                node.prev = prev;
                 return;
             }
             prev = current;
-            current = (*current).next;
+            current = current_node.as_ref().next;
         }
 
-        // We looped through the whole list and the item is bigger or equal
+        // We looped through the whole list and the nodes data is bigger or equal
         // than everything we found up to now.
         // Insert at the end. Since we checked before that the list isn't empty,
         // tail always has a value.
-        (*item).prev = self.tail;
-        (*item).next = null_mut();
-        (*self.tail).next = item;
-        self.tail = item;
+        node.prev = self.tail;
+        node.next = None;
+        self.tail.as_mut().unwrap().as_mut().next = Some(node.into());
+        self.tail = Some(node.into());
     }
 
-    /// Returns the first item in the linked list without removing it from the list
+    /// Returns the first node in the linked list without removing it from the list
     /// The function is only safe as long as valid pointers are stored inside
     /// the linked list.
     /// The returned pointer is only guaranteed to be valid as long as the list
     /// is not mutated
-    pub fn peek_first(&self) -> *mut ListNode<T> {
-        self.head
+    pub fn peek_first(&self) -> Option<&mut ListNode<T>> {
+        // Safety: When the node was inserted it was promised that it is alive
+        // until it gets removed from the list.
+        // The returned node has a pointer which constrains it to the lifetime
+        // of the list. This is ok, since the Node is supposed to outlive
+        // its insertion in the list.
+        unsafe {
+            self.head
+                .map(|mut node| &mut *(node.as_mut() as *mut ListNode<T>))
```

Comment:
nit: can't this just be 
```suggestion
                .map(|node| node.as_mut())
```
or even just
```suggestion
                .map(NonNull::as_mut)
```

Reply:
No. I think that is what I had first. However it doesn't believe the lifetime of the reference is ok:

```
cannot return value referencing function parameter `node`
returns a value referencing data owned by the current function
```